test/Constraints/diagnostics.swift

```diff
@@ -300,6 +300,59 @@ func rdar21784170() {
   (Array.init as (Double...) -> Array<Double>)(initial as (Double, Double)) // expected-error {{cannot convert value of type '(Double, Double)' to expected argument type 'Double'}}
 }
 
+// Diagnose passing an array in lieu of variadic parameters
+func variadic(_ x: Int...) {}
+func variadicArrays(_ x: [Int]...) {}
+func variadicAny(_ x: Any...) {}
+struct HasVariadicSubscript {
+  subscript(_ x: Int...) -> Int {
+    get { 0 }
+  }
+}
+let foo = HasVariadicSubscript()
+
+let array = [1,2,3]
+let arrayWithOtherEltType = ["hello", "world"]
+
+variadic(array) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}} {{10-10=#variadic(}} {{15-15=)}}
+variadic([1,2,3]) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}} {{10-10=#variadic(}} {{17-17=)}}
+// expected-note@-2 {{remove brackets to pass array elements directly}} {{10-11=}} {{16-17=}}
+variadic([1,2,3,]) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}} {{10-10=#variadic(}} {{18-18=)}}
+// expected-note@-2 {{remove brackets to pass array elements directly}} {{10-11=}} {{16-17=}} {{17-18=}}
+variadic(0, array, 4) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}}
+variadic(0, [1,2,3], 4) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}}
+// expected-note@-2 {{remove brackets to pass array elements directly}} {{13-14=}} {{19-20=}}
+variadic(0, [1,2,3,], 4) // expected-error {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
+// expected-note@-1 {{wrap the expression in '#variadic(...)' to pass array elements as variadic arguments}}
+// expected-note@-2 {{remove brackets to pass array elements directly}} {{13-14=}} {{19-20=}} {{20-21=}}
+variadic(arrayWithOtherEltType) // expected-error {{cannot convert value of type '[String]' to expected argument type 'Int'}}
+variadic(1, arrayWithOtherEltType) // expected-error {{cannot convert value of type '[String]' to expected argument type 'Int'}}
+variadic(["hello", "world"]) // expected-error 2 {{cannot convert value of type 'String' to expected element type 'Int'}}
+// expected-error@-1 {{cannot pass an array of type '[Int]' as variadic arguments of type 'Int'}}
```

Comment:
Hmm, that's unfortunate. The good news at least is that it's not a specific problem with the fix you've introduced, it's a more general problem when we have multiple contextual fixes in play. I've filed a bug to track that: https://bugs.swift.org/browse/SR-11104.

Reply:
Thanks for taking the time to file this!